frame/composable-traits/src/dex.rs

```diff
@@ -258,26 +260,47 @@ pub trait DexRouter<AccountId, AssetId, PoolId, Balance, MaxHops> {
 		asset_pair: CurrencyPair<AssetId>,
 		route: Option<BoundedVec<PoolId, MaxHops>>,
 	) -> Result<(), DispatchError>;
-	/// If route exist return `Some(Vec<PoolId>)`, else `None`.
-	fn get_route(asset_pair: CurrencyPair<AssetId>) -> Option<Vec<PoolId>>;
-	/// Exchange `dx` of `base` asset of `asset_pair` with associated route.
+	/// If route exist return `Some((Vec<PoolId>, bool))`, else `None`.
+	/// boolean in pair indicates if route needs to be used in reversed direction.
+	fn get_route(asset_pair: CurrencyPair<AssetId>) -> Option<(Vec<PoolId>, bool)>;
+	/// Exchange `amount` of `quote` asset of `asset_pair` with associated route.
 	fn exchange(
 		who: &AccountId,
 		asset_pair: CurrencyPair<AssetId>,
-		dx: Balance,
+		amount: Balance,
+		min_receive: Balance,
 	) -> Result<Balance, DispatchError>;
-	/// Sell `amount` of `base` asset of asset_pair with associated route.
+	/// Sell `amount` of `quote` asset of asset_pair with associated route.
 	fn sell(
 		who: &AccountId,
 		asset_pair: CurrencyPair<AssetId>,
 		amount: Balance,
+		min_receive: Balance,
 	) -> Result<Balance, DispatchError>;
 	/// Buy `amount` of `quote` asset of asset_pair with associated route.
 	fn buy(
 		who: &AccountId,
 		asset_pair: CurrencyPair<AssetId>,
 		amount: Balance,
+		min_receive: Balance,
 	) -> Result<Balance, DispatchError>;
+	/// Add liquidity to underlying dex pool. Works only for route with single pool.
+	fn add_liquidity(
+		who: &AccountId,
+		asset_pair: CurrencyPair<AssetId>,
+		base_amount: Balance,
+		quote_amount: Balance,
+		min_mint_amount: Balance,
+		keep_alive: bool,
+	) -> Result<(), DispatchError>;
+	/// Remove liquidity from underlying dex pool. Works only for route with single pool.
+	fn remove_liquidity(
+		who: &AccountId,
+		asset_pair: CurrencyPair<AssetId>,
+		lp_amount: Balance,
+		min_base_amount: Balance,
+		min_quote_amount: Balance,
+	) -> Result<(), DispatchError>;
```

Comment:
Does not serve any purpose? What if another prtocol has a dependency to Amm and I want to use either Pablo or Dex router? Isn't that a purpose? This duplicated code make it way more complicated to maintain than having the same interface implemented by multiple pallet :). Could we try? For the update_route it's out of Amm scope so it's fine to have a Router interface as currently defined

Reply:
Okay, so I tried changing it but it requires some more modification like in Amm we don't have min_receive like parameters on buy/sell but in current implementation of DexRouter I have added it. So I think first lets complete this PR and then delete dex pallets except pablo(I think we can delete them now) and refactor DexRouter to have Amm trait and it will also require some small modification to Pablo's AMM .
@hussein-aitlahcen agree?

Reply:
It's actually straight forward, the pallet must implements min_receive anyway for user to specify slippage. We currently default to 0 i believe

Reply:
Yes but better to remove other dex first so that I don't have to update their code base, and thus keep PR size in limit so reviews are faster and less error prone.

Reply:
Ok